### PR TITLE
Improve relevance scoring for titles and object-name matches in search results

### LINKED ISSUE #12391: [search] issues with the new HTML search algorithm
I don't really have a good title for this issue, but the CPython issue speaks for itself. It appears we are still misranking titles in some cases (the 3.11 doc is generated by 7.2.6 but the 3.12 is generated by 7.3.7).

cc @wlach @jayaddison 

From https://github.com/python/cpython/issues/119423#issue-2311184678

> (I initially posted this in the [pythondotorg](https://github.com/python/pythondotorg/issues/2452) repo - where I was directed here. Copying the description from there verbatim)
> 
> **Describe the bug** Searching for library modules on 3.12 returns first a list of "What's new in <version>" pages instead of the searched for module.
> 
> **To Reproduce** Steps to reproduce the behavior:
> 
>     1. Go to [docs.python.org/3.12/index.html](https://docs.python.org/3.12/index.html)
> 
>     2. Click on search box in top right corner, enter e.g. "asyncio" and hit enter
> 
>     3. [Scroll through results](https://docs.python.org/3.12/search.html?q=asyncio)
> 
>     4. See that the first result is "History and License" followed by 10 (!) "What’s New In Python <version>" results. The 11th result is the actual asyncio module.
> 
> 
> **Expected behavior** I would expect the asyncio module page to be the first result, or at the very least to appear very close to the top. This has been the case in previous versions of the website search.
> 
> **Screenshots** (the list goes on a few more entries below the screenshot) image
> 
> Compare with search from 3.11: image
> 
> **Desktop (please complete the following information):**
> 
>     * OS: MacOS
> 
>     * Browser Chrome
> 
>     * Version 123
> 
> 
> **Additional context** This happens for other topics too, not only modules.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Resolve some sub-optimal search result ranking/scoring reported by #12391.

### Detail
- [x] Add an example project to the JavaScript tests that exhibits the sub-optimal search result behaviour.
- [x] Add JavaScript test coverage to assert on the expected, improved relative ordering of query results.
- [x] Implement changes to the indexing/query algorithms to improve the query results without regressing other JavaScript search tests.
  - [x] Merged to fbb62cfda9600d2caa4d6f097745d59cea3a431b from #12393 (thanks @wlach!) and then added some suggested refactoring to that. 

### Relates
- Resolves #12391.

### CODE REVIEW

**Comment by jayaddison** [comment]
Commit cb0f6e7ffe5ec3afe8beebfeeba24b50451e7971 -- regenerating a search index file from scratch -- seems to be have been necessary because I had a stale `_build` directory on my local machine for the relevant input project for the fixture.

That seems like a bug; re-using an existing `_build` directory should be valid behaviour and should produce the same search index output as a fresh build.  I'll file that as a separate issue within the next few days.

**Comment by jayaddison** [comment]
I've a slight preference for #12047 to be merged before this, to make the code and diff history easier to follow, if+when either of them are considered ready.

File: sphinx/themes/basic/static/searchtools.js [hunk 1]
```diff
@@ -39,6 +39,7 @@ if (typeof Scorer === "undefined") {
     objPrioDefault: 0,
 
     // query found in title
+    mainTitle: 16,
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js]: I've been thinking about this a bit-- technically this breaks backwards compatibility if someone defined a "custom" scorer. I think we might be better off doing something like:

```js
if (typeof Scorer === "undefined") {
  Scorer = {};
}
Scorer = {
   mainTitle: 16,
   ...
   ...Scorer,
}
```

That way we can add new properties to the class and if the user over-rode it we'd fall back to the new default values.

File: sphinx/themes/basic/static/searchtools.js [hunk 2]
```diff
@@ -328,7 +329,7 @@ const Search = {
     for (const [title, foundTitles] of Object.entries(allTitles)) {
       if (title.toLowerCase().trim().includes(queryLower) && (queryLower.length >= title.length/2)) {
         for (const [file, id] of foundTitles) {
-          let score = Math.round(100 * queryLower.length / title.length)
+          let score = Math.round((titles[file] !== title ? Scorer.title : Scorer.mainTitle) * queryLower.length / title.length)
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js]: I think I preferred my approach of creating a temporary called `isMainTitle` and then reusing it both here and below, but it's not that big a deal.

File: tests/js/searchtools.js [hunk 3]
```diff
@@ -94,6 +109,25 @@ describe('Basic html theme search', function() {
 
   });
 
+  describe('search result ranking', function() {
+
+    it('should score an object-name match above a page-title match', function() {
```
Comment (wlach) [inline tests/js/searchtools.js]: ... and a page-title match above above a subtitle match? Maybe easier to just say "it should rank things in the appropriate order" and then explain the details in a comment (maybe with a reference to #12391 for context)

**Review by wlach — APPROVED** [review]
Looks good to me! Some minor optional comments.

File: sphinx/themes/basic/static/searchtools.js (hunk 1 above)
Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: Hm.  That is true; but could that breakage be avoided by setting the original `Scorer` class as the `__proto__` (prototype) for the custom class?  Inheritance-wise, I _think_ that would be the safe way to overload the values -- and perhaps a valid migration path for any sites that had simply defined their own without using inheritance?

File: sphinx/themes/basic/static/searchtools.js (hunk 2 above)
Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: Let me see if I can find a middle-ground (that might also resolve the overloads-backwards-compat concern).

Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: Actually, nope - I think I agree that your approach is simpler :)  I'll revert to that.

Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: I'll push one more variation, to see what you think.  I want to make it as clear as possible that there are essentially two changes we're making here: (1) changing the base score from `100` to `Scorer.title`, and (2) adding a boost for main document titles.

File: tests/js/searchtools.js (hunk 3 above)
Comment (jayaddison) [reply tests/js/searchtools.js]: If possible I'd like to move in a more granular direction with these tests; breaking this into two:

  * Main document title should rank above section title (for equivalent/similar string matches).
  * Programming language domain object should rank above main document title (again for equivalent/similar string matches).

The problem with larger lists of "these things should rank in this order" is that it could lead to difficult-to-comprehend logic in the code that is more of an attempt by developers to keep the tests passing than to retain a set of sensible rules.  It does get tricky over time with either approach, but I think it's useful to be able to identify what specific rankings are broken (and when).

Links to GitHub issues and external resources can be a useful resource sometimes (like the duplicate-result test fixture data / TODO note recently) -- if we can though, I'd prefer to keep the code and tests self-contained and self-documenting.

Comment (wlach) [reply tests/js/searchtools.js]: Okay! To be honest I don't have super strong opinions here as long as we have a test of some kind.

**Comment by jayaddison** [comment]
This should be ready for further review / merge; I've no changes planned on this branch.

File: tests/js/searchtools.js (hunk 3 above)
Comment (jayaddison) [reply tests/js/searchtools.js]: > ... and a page-title match above above a subtitle match? [ ... snip ... ]

That's now in place more clearly, by splitting it out into a separate test case (and indirectly that also helped to find bugs in the `checkRanking` test utility function..).

File: sphinx/themes/basic/static/searchtools.js [hunk 4]
```diff
@@ -328,13 +328,14 @@ const Search = {
     for (const [title, foundTitles] of Object.entries(allTitles)) {
       if (title.toLowerCase().trim().includes(queryLower) && (queryLower.length >= title.length/2)) {
         for (const [file, id] of foundTitles) {
-          let score = Math.round(100 * queryLower.length / title.length)
+          let score = Math.round(Scorer.title * queryLower.length / title.length);
+          let boost = titles[file] === title ? 1 : 0;  // add a boost for document titles
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js]: On second look, could these be declared as `const`?

```suggestion
          const score = Math.round(Scorer.title * queryLower.length / title.length);
          const boost = titles[file] === title ? 1 : 0;  // add a small boost for document titles
```

File: tests/js/searchtools.js [hunk 5]
```diff
@@ -85,6 +102,38 @@ describe('Basic html theme search', function() {
...
+        ['relevance', 'Relevance', ''],  /* main title */
+      ];
+
+      searchParameters = Search._parseQuery('relevance');
+      results = Search._performSearch(...searchParameters);
+
+      checkRanking(expectedRanking, results);
+    });
+
+    it('should score an main-title match above a subheading-title match', function() {
```
Comment (wlach) [inline tests/js/searchtools.js]:
```suggestion
    it('should score a main-title match above a subheading-title match', function() {
```

File: tests/js/searchtools.js [hunk 6]
```diff
@@ -7,6 +7,23 @@ describe('Basic html theme search', function() {
       return req.responseText;
   }
 
+  function checkRanking(expectedRanking, results) {
+    let [nextExpected, ...remainingItems] = expectedRanking;
```
Comment (wlach) [inline tests/js/searchtools.js:11]: Apologies if I didn't notice this earlier, but I'm a little unclear on why this function isn't expressed as something like:

```js
expect(results.map(r=>r.slice(0, 3)).toEqual(expectedRanking)
```

I think that would likely give a clearer error message in the event that the output deviated from what was expected.

**Review by wlach — COMMENTED** [review]
I have a couple of extra comments but in general this LGTM. I think this will be a nice incremental improvement. @picnixz may you could take a look too (and merge when you think it ready)?

File: CHANGES.rst [hunk 7]
```diff
@@ -92,6 +92,11 @@ Bugs fixed
 * #11961: Omit anchor references from document title entries in the search index,
   removing duplication of search results.
   Patch by James Addison.
+* #12391: Adjust scoring of matches during HTML search so that document main
+  titles tend to rank more highly than subsection titles, and also to provide
+  a gain to matches on the name of programming domain objects relative to
+  title/subtitle matches.
+  Patch by James Addison and Will Lachance.
```
Comment (picnixz) [inline CHANGES.rst]:
```suggestion
* #12391: Adjust scoring of matches during HTML search so that main
  titles tend to rank higher than subsection titles. In addition, 
  boost matches on the name of programming domain objects relative
  to title/subtitle matches.
  Patch by James Addison and Will Lachance.
```

File: sphinx/themes/basic/static/searchtools.js (hunk 4 above)
Comment (picnixz) [reply sphinx/themes/basic/static/searchtools.js]: I thinkg it's better using a const as well. But on a second thought, I'm wondering whether a +1 is sufficient.

File: tests/js/roots/titles/relevance.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+class Example:
+    """Example class"""
+    num_attribute = 5
+    text_attribute = "string"
```
Comment (picnixz) [inline tests/js/roots/titles/relevance.py:4]: Add an attribute that has the same name as a title as well.

File: tests/js/searchtools.js (hunk 6 above)
Comment (picnixz) [reply tests/js/searchtools.js:11]: I concur.

**Review by picnixz — COMMENTED** [review]
It'd be good if we have a more complete example where you have a lot of multiple matches of the same kind. Does it cover the issue with the asyncio module that we described?

File: tests/js/searchtools.js (hunk 6 above)
Comment (jayaddison) [reply tests/js/searchtools.js:11]: Good question; it was in fact intentional that I decided to begin by checking the relative positions of expected results, instead of for a precise ordering within the search results.  To explain:

If we begin introducing more of these kind of `checkRanking` test cases, and fixing/adjusting the search code over time, then precise-order assertions could become quite fragile; we'd likely need to update multiple tests during every change.

That's not a situation I'd like: I think our tests should state our general, reasonable expectations about what kind of results should rank higher than others -- without proscribing exactly where they'll appear.

Tests written that way should allow more space for implementation/search fixes, while still describing the expected behaviour of the search engine.

In the case that we reach a change/conflict where one of these tests _has_ to be updated, that's an indicator that we might need to debate the test cases we've selected and possibly drop or modify old ones -- but the intent is for that to be a rare (and discussion-worthy) scenario.

I'm not sure whether I explained that philosophy very well - let me know if that makes sense, or if I should try to find other ways to phrase it.

Edit: typo fixup

**Comment by jayaddison** [comment]
> It'd be good if we have a more complete example where you have a lot of multiple matches of the same kind. Does it cover the issue with the asyncio module that we described?

Yep, the thinking here was to replicate the `asyncio` relevance ordering problem using a minimal test case, and then to adjust the code to fix it; attempting to apply (and demonstrate) a Test-Driven-Development approach to search ranking fixups.  I'll investigate expanding the test fixture data to add more results.

File: tests/js/searchtools.js (hunk 6 above)
Comment (wlach) [reply tests/js/searchtools.js:11]: Hmm, I see. I have to confess I'm not sure if I'm totally convinced: at least at this point, the test cases are simple enough that this feels a bit like overengineering. I'd personally do the simpler approach for now, then if we find we're accumulating a lot of different examples and it's becoming difficult to reason about the effect of changes at a glance, we could add something like this.

If you're determined to use this approach out of the gate, I'd add a comment explaining the reasoning.

**Comment by wlach** [comment]
> It'd be good if we have a more complete example where you have a lot of multiple matches of the same kind. Does it cover the issue with the asyncio module that we described?

This PR uses a similar approach to what was described/shown in https://github.com/sphinx-doc/sphinx/pull/12441#pullrequestreview-2165221022 so it should.

However, it would be good to another test before landing to be sure. I tested there by checking out the [cpython repository](https://github.com/python/cpython/) and regenerating the `Doc/` directory using a virtualenv with my development version of Sphinx.

File: sphinx/themes/basic/static/searchtools.js (hunk 4 above)
Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: Previously a title and subsection title with the same text would have equal scores, leaving their relative ranking undefined.

Any positive value here should have the effect of elevating the main-document titles above same-named subsection titles in the search results.

A single-integer increment is used because ideally we don't want the main document titles to move up in the rankings 'too much' and overtake other matches.  That is possible, though, especially given that some scores are fractional.  So I have the opposite worry: that +1 might be too much.

File: tests/js/searchtools.js (hunk 6 above)
Comment (jayaddison) [reply tests/js/searchtools.js:11]: I think I'm reasonable determined.  It's intended to solve the today-problem of "does this fix the result ordering bug?", the middle-term problems of "describe the search ranking logic we use, and try to catch breaking changes", and the long-ish term social problem of "project `foo` wants document `bar` to rank higher; why isn't it and what are the options to address that?".

Arguably folks could use a framework like this to self-check their projects' search result quality during continuous integration and so on.  But _that_ is definitely overengineering at this point :)

So in short: yep, I'll document these ideas in the test suite (and if they're amended or removed in future as other approaches are followed, that's fine too).

File: sphinx/themes/basic/static/searchtools.js (hunk 4 above)
Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: (a good way to figure these out could be to develop counterexamples and add test cases for them)

File: tests/js/searchtools.js [hunk 9]
```diff
@@ -85,6 +102,52 @@ describe('Basic html theme search', function() {
...
+     * [1] - https://github.com/sphinx-doc/sphinx.git/
+     */
+
+    it('should score an object-name match above a page-title match', function() {
+      eval(loadFixture("titles/searchindex.js"));
+
+      expectedRanking = [
+        ['index', 'relevance', '#module-relevance'],  /* py:module documentation */
+        ['relevance', 'Relevance', ''],  /* main title */
+      ];
```
Comment (jayaddison) [inline tests/js/searchtools.js:127]: I'm questioning my own test case a bit here: is it really true that an object name should match about a main document title in the general case?

File: tests/js/searchtools.js (hunk 6 above)
Comment (wlach) [reply tests/js/searchtools.js:11]: Ok! I don't feel strongly enough about this myself to continue arguing. Good enough for now, safe enough to try. 😄

File: tests/js/searchtools.js (hunk 9 above)
Comment (jayaddison) [reply tests/js/searchtools.js:127]: Questioning this further: is the `other.Example.relevance` attribute more relevant to the search query `relevance` than the document with title `Relevance`?  I would probably suggest not (and perhaps I should have chosen a different word to use for these test cases).

Comment (jayaddison) [reply tests/js/searchtools.js:127]: Yep, after some more thought I think I'd suggest these rules -- all in the context of exact-match names (partial matches are worth considering, but later):

  * A ['high priority' object match](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/themes/basic/static/searchtools.js#L34) (such as a Python module name) should rank above a main document title match.
  * A main document title match should rank above a normal priority object match (such as a Python class attribute).

Achieving this does require one more change to the search code; I've prototyped that locally to produce the results displayed below:

image

Comment (jayaddison) [reply tests/js/searchtools.js:127]: > Achieving this does require one more change to the search code; I've prototyped that locally to produce the results displayed below:

My mistake there; no code change is required.  I've updated the test suite in this branch to include the suggested rules, though.

**Comment by AA-Turner** [comment]
@jayaddison are you happy with this // ready to review & merge?

A

**Comment by jayaddison** [comment]
@AA-Turner yep, I think this is ready.

**Comment by jayaddison** [comment]
Thank you @AA-Turner!
